resources/puppetlabs/puppetdb/pql/pql-grammar-experimental.ebnf

```diff
@@ -0,0 +1,139 @@
+
+(* Static list of entity types *)
+<entity> = 'facts' |
+           'reports' |
+           'catalogs' |
+           'resources' |
+           'nodes' |
+           'environments' |
+           'factsets' |
+           'fact_paths' |
+           'fact_contents' |
+           'edges' |
+           'events';
+
+(* Expressions
+
+   This structure preserves precedence rules after parsing, so for reference the number
+   after expr is the order of precedence. One should be careful when modifying
+   this structure, and ensure that precedence rules continue to be kept. *)
+<expression> = expr4;
+expr4   = expr3 { <whitespace>, or, <whitespace>, expr4 };
+expr3   = expr2 { <whitespace>, and, <whitespace>, expr3 };
+expr2   = ( not, [<whitespace>], expr2 ) | expr1;
+<expr1> = ( <lparens>, expression, <rparens> ) | condexpression | subquery;
+
+(* Implicit subqueries *)
+subquery = entity, [<whitespace>], where;
+
+(* Conditional expression *)
+condexpression    = (condexpregexp | condexpnumber | condexpmatch | condexpsubquery | condexpnull);
+<condexpregexp>   = field, [<whitespace>], condregexp, [<whitespace>], valueregexp;
+<condexpnumber>   = field, [<whitespace>], condnumber, [<whitespace>], valuenumber;
+<condexpmatch>    = field, [<whitespace>], condmatch, [<whitespace>], valuematch;
+<condexpsubquery> = (field | groupedfieldlist), [<whitespace>], condsubquery, [<whitespace>], query;
+<condexpnull>     = field, [<whitespace>], condnull, [<whitespace>], boolean;
+
+(* Conditional expression parts *)
+groupedfieldlist = <lparens>, [<whitespace>], fieldlist, [<whitespace>], <rparens>;
+<fieldlist>      = field, [ [<whitespace>], <','>, [<whitespace>], fieldlist ];
+<field>          = #'[a-zA-Z_]+';
+
+<condregexp>    = '~';
+<condnumber>    = '>=' | '<=' | '<' | '>';
+<condmatch>     = '=';
+<condsubquery>  = 'in';
+<condnull>      = 'null?';
+
+<valueregexp>   = regexp;
+<valuenumber>   = integer | real;
+<valuematch>    = string | integer | real | boolean;
+
+(* Boolean operators *)
+<and> = <'and'>;
+<or>  = <'or'>;
+not   = <'!'>;
+
+(* Regexp strings *)
+regexp = <regexpquote>, stringwithoutregexpquote, <regexpquote>
+<stringwithoutregexpquote> = #'(?:[^/\\]|\\.)*';
+<regexpquote> = "/"
+
+(* Basic strings
+
+   Here we differentiate between double quoted and single quoted
+   strings so we can determine how to strip escapes later on. *)
+<string> = dqstring | sqstring;
+dqstring = (<doublequote>, stringwithoutdoublequotes, <doublequote>);
+<stringwithoutdoublequotes> = #'(?:[^"\\]|\\.)*';
+<doublequote> = "\"";
+
+sqstring = (<singlequote>, stringwithoutsinglequotes, <singlequote>);
+<stringwithoutsinglequotes> = #'(?:[^\'\\]|\\.)*';
+<singlequote> = "'";
+
+(* Parens *)
+<lparens> = '(';
+<rparens> = ')';
+
+(* Braces *)
+<lbrace> = '{';
+<rbrace> = '}';
+
+(* Bracket *)
+<lbracket> = '[';
+<rbracket> = ']';
+
+(* Booleans *)
+boolean = true | false;
+true = <"true">;
+false = <"false">;
+
+(* Numbers *)
+integer    = [negative], digits;
```

Comment:
I've fixed support for the + modifier no problem.

But there is a gotcha with support integer e notation, in that the Java Integer parseInt support doesn't support it. Hilarious really, but then I started to realise something ... what is 1E-5? Its a float ... but ... 1E5, its an integer. By trying to 'figure out' its still an integer and doing something magical, this can create some confusion, or we could just presume anything with E notation is a real. 

```
puppetlabs.puppetdb.core=> (Double. "1e-1")
0.1
puppetlabs.puppetdb.core=> (Integer. "1e-1")
NumberFormatException For input string: "1e-1"  java.lang.NumberFormatException.forInputString (NumberFormatException.java:65)

puppetlabs.puppetdb.core=> (Double. "1e1")
10.0
puppetlabs.puppetdb.core=> (Integer. "1e1")
NumberFormatException For input string: "1e1"  java.lang.NumberFormatException.forInputString (NumberFormatException.java:65)
```